### ISSUE TITLE
Lia confused by dependently typed hypotheses

### ISSUE THREAD

**Issue by maximedenes** [coq/coq#7886]
In the following context:

```coq
Require Import Lia.

Axiom decompose_nat : nat -> nat -> nat.
Axiom inleft : forall {P}, {m : nat & P m} -> nat.
Axiom foo : nat.

Goal forall (x x0 : nat)
  (e : 0 = x0 + S x)
  (H : decompose_nat x 0 = inleft (existT (fun m : nat => 0 = m + S x) x0 e))
  (x1 : nat)
  (e0 : 0 = x1 + S (S x))
  (H1 : decompose_nat (S x) 0 = inleft (existT (fun m : nat => 0 = m + S (S x)) x1 e0)),
    False.
Proof.
intros.
```

`lia` fails while `clear H H1; lia` succeeds.

`omega` doesn't have this problem.

I found many occurrences of this while porting VST to `lia`.

**Comment by maximedenes** [comment]
@fajb do you have an idea where this bug could come from?

**Comment by gares** [comment]
Looks like it's time for a `Set Debug micromega` https://github.com/coq/coq/blob/master/plugins/micromega/coq_micromega.ml#L32

**Comment by maximedenes** [comment]
Indeed!

**Comment by fajb** [comment]
This is actually a limitation of zify: the hypotheses e and e0 are not mapped to Z because they are used in dependent hypotheses.

Maybe, @letouzey knows how to do it? 

Removing dependent hypotheses would not be satisfactory as the proof may need them...

**Comment by maximedenes** [comment]
How does `omega` succeed, then?

**Comment by JasonGross** [comment]
It seems like a simple solution would be to duplicate hypotheses (with `pose proof`) in cases where `rewrite` fails, and use a "marker" (e.g., `Definition marker {T} (x : T) := x.`) to mark hypotheses which should not be duplicated (because they've already been duplicated).

Thinking a bit more, though, since `zify` is internal to these decision procedures, and all of them are supposed to either finish the proof or fail, it should be safe to remove dependent hypotheses, right?

**Comment by JasonGross** [comment]
@maximedenes I suspect `omega` ignores the hypotheses it can't understand.  Since it can natively understand `nat` hypotheses, it can use them directly, without caring about the fact that other hypotheses depend on them.

**Comment by Zimmi48** [comment]
> How does omega succeed, then?

Indeed, `omega` does understand `nat` hypotheses natively, whereas `romega` and `lia` depend on `zify`. The example you gave is also one where `romega` fails.

On this simpler example:

```
Require Import Lia Omega ROmega.

Axiom decompose_nat : nat -> nat -> nat.
Axiom inleft : forall {P}, {m : nat & P m} -> nat.

Goal forall (x x0 : nat) (e : 0 = x0 + S x),
  decompose_nat x 0 <> inleft (existT (fun m : nat => 0 = m + S x) x0 e).
Proof.
intros.
```

`zify` gives:

```
1 subgoal
x, x0 : nat
e : 0 = x0 + S x
H : (0 <= Z.of_nat (decompose_nat x 0))%Z
H0 : (0 <=
      Z.of_nat
        (inleft
           (existT
              (fun m : nat => 0%nat = (m + S x)%nat) x0
              e)))%Z
______________________________________(1/1)
Z.of_nat (decompose_nat x 0) <>
Z.of_nat
  (inleft (existT (fun m : nat => 0 = m + S x) x0 e))
```

Asking `zify` to duplicate dependent hypotheses like suggested by @JasonGross to only transfer one copy of them seems like a good idea:

```
1 subgoal
x, x0 : nat
e : 0 = x0 + S x
e' : 0%Z = (Z.of_nat x0 + Z.succ (Z.of_nat x))%Z
H : (0 <= Z.of_nat x0)%Z
H0 : (0 <= Z.of_nat x)%Z
H0 : (0 <=
      Z.of_nat
        (inleft
           (existT
              (fun m : nat => 0%nat = (m + S x)%nat) x0
              e)))%Z
______________________________________(1/1)
Z.of_nat (decompose_nat x 0) <>
Z.of_nat
  (inleft (existT (fun m : nat => 0 = m + S x) x0 e))
```

**Comment by maximedenes** [comment]
Now fixed as part of #7878.

**Comment by Blaisorblade** [comment]
For the record; "now fixed" means "will be fixed when #7878 is merged". Is there any chance to merge the fixes before the removal of omega & romega?

**Comment by maximedenes** [comment]
> Is there any chance to merge the fixes before the removal of omega & romega?

Yes, I'll try to extract the fix.

**Comment by maximedenes** [comment]
> Yes, I'll try to extract the fix.

I remember now: the fix is incorrect. It requires to iter over a context while generating new hypotheses on the fly, which is a mess to do in Ltac1.

This issue will be fixed by #9856 instead.
